workshop/content/030_basic_content/075_serverless/_index.md

```diff
@@ -5,7 +5,264 @@ weight: 75
+
+{{% expand "Troubleshooting" %}}
+If the experiment fails to run, review the **Troubleshooting** heading at the bottom of the [**FIS SSM Start Automation Setup**]({{< ref "030_basic_content/030_basic_experiment/" >}}) section. Note that there is a possibility that SSM automation will not start at all. If this is the case, especially if you have manually created any of the resources, double check that you are specifying full ARNs rather than just names in the FIS template.
+{{% /expand %}}
+
+
+In the FIS window select the "Timeline" tab and hit refresh every minute or so. You should see the experiment progressing through the individual states with green indicating finished steps, blue indicating in-progress steps, and grey signifying steps yet to be started:
+
+{{< img "timeline.png" "Experiment timeline" >}}
+
+At the same time, watch the curl output in the CloudShell window. As the experiment transitions from one step to the next you should see the output change, first to increase the latency:
+
+```
+...
+Hello from Lambda! - 200 - 0.541716
+Hello from Lambda! - 200 - 0.513623
+Hello from Lambda! - 200 - 0.546924
+...
+```
+
+then for the latency to return to normal but the response code changing to 404:
+
+```
+...
+Hello from Lambda! - 404 - 0.113380
+Hello from Lambda! - 404 - 0.274236
+Hello from Lambda! - 404 - 0.136305
+...
+```
+
+and finally changing to an error message indicating an exception has occured during code execution:
+
+```
+...
+{"message": "Internal server error"} - 502 - 0.215665
+{"message": "Internal server error"} - 502 - 0.113820
+{"message": "Internal server error"} - 502 - 0.163391
+...
+```
+
+before returning to normal at the end of the experiment.
+
+Congratulations for completing this lab! In this lab you walked through running a multi-step experiment changed an SSM Parameter Store parameter and injected faults into a Lambda function. 
+
+
+## Learning and improving
+
+The setup we've shown here provides failure modes similar to those availabel for instances and containers. It also has various problems that you can experiment with and use for ideation on how to customize your own serverless fault injection libraries:
+
+* If you stop the FIS experiment prematurely, the parameter will not be reset to a non-impacting configuration. To address this you could add a `NoFaultParameterValue` parameter to the SSM document / FIS template and add an `onError` / `onCancel` path to the SSM document. You could even read the parameter at the start of the SSM document run, but consider concurrency implications if anyother experiment is also changing the parameter.
+* If you use the `status-code` error, the Lambda code still executes but reports a failure when no failure occurred. Similarly, in the current implementation an exception occurs before executing user code but could be moved to occur after user code. What impact would the mis-reporting of errors have on error handling in downstream systems?
```

Comment:
That needs to be fixed into my code. Can you open an issue please?

Reply:
Remember how I asked if I could be included in writing the "go code" for the lambda extensions? It was because of this - my opinion that structurally we should have pre/post hooks and be able to plug in multiple events into each using the parameter.

Issue created https://github.com/adhorn/aws-lambda-chaos-injection/issues/32

Reply:
Remember how I asked if I could be included in writing the "go code" for the lambda extensions? It was because of this - my opinion that structurally we should have pre/post hooks and be able to plug in multiple events into each using the parameter.

Issue created https://github.com/adhorn/aws-lambda-chaos-injection/issues/32